tests/ext_data_control_v1.cpp

```diff
@@ -0,0 +1,600 @@
+
+    ExtDataControlClient(Server& the_server) :
+        Client{the_server},
+        data_control_manager{bind_if_supported<ext_data_control_manager_v1>(AnyVersion)},
+        data_control_device{ext_data_control_manager_v1_get_data_device(data_control_manager, seat())},
+        source_data_control_source{ext_data_control_manager_v1_create_data_source(data_control_manager)}
+    {
+    }
+
+    void as_sink()
+    {
+        receiving_pipe.emplace();
+
+        ext_data_control_device_v1_set_user_data(data_control_device, this);
+        // Set up the control device listener.
+        // The next roundtrip should cause `selection` to be called with a null
+        // id once the the control device is created on the server side.
+        ext_data_control_device_v1_add_listener(data_control_device, &sink_device_listener, this);
+
+        // On the next roundtrip, the sink will receive notification of offer
+        // on data control device -> Create offer and store it
+        // Receive notification of offer selection/primary selection on control
+        // device -> Send `send()` event with fd and mime type
+    }
+
+    auto try_read() -> std::string
+    {
+        char buf[128];
+        auto const read_chars = read(receiving_pipe->sink, buf, sizeof(buf));
+        received_message = std::string{buf, static_cast<size_t>(read_chars)};
+        return *received_message;
+    }
+
+    struct SourceOptions
+    {
+        SelectionType selection = SelectionType::normal;
+        std::optional<std::string> message = {};
+        std::function<void()> when_content_sent = []{};
+    };
+
+    void as_source(SourceOptions&& options)
+    {
+        if(options.message)
+            received_message = options.message;
+
+        this->when_content_sent = options.when_content_sent;
+
+        ext_data_control_source_v1_add_listener(
+            source_data_control_source, &source_source_listener, this);
+
+        // Offer plaintext
+        ext_data_control_source_v1_offer(source_data_control_source, test_mime_type);
+
+        switch (options.selection)
+        {
+        case SelectionType::normal:
+
+            // Set selection
+            // For all current and future clients, this should notify them that
+            // this source client is the active source when they create their
+            // devices via the `selection()` event.
+            ext_data_control_device_v1_set_selection(data_control_device, source_data_control_source);
+            break;
+        case SelectionType::primary:
+            ext_data_control_device_v1_set_primary_selection(data_control_device, source_data_control_source);
+            break;
+        }
+    }
+};
+
+struct ExtDataControlV1Test : public wlcs::StartedInProcessServer
+{
+};
+}
+
+TEST_F(ExtDataControlV1Test, client_copies_into_clipboard_that_started_after_it)
+{
+    auto copying_client = ExtDataControlClient{the_server()};
+    auto clipboard_client = ExtDataControlClient{the_server()};
```

Comment:
This isn't idiomatic C++

```suggestion
    ExtDataControlClient copying_client{the_server()};
    ExtDataControlClient clipboard_client{the_server()};
```